bin/utest/mmap.c

```diff
@@ -94,15 +94,26 @@ static void munmap_good(void) {
   assert(result == 0);
 }
 
-/* Don't call this function in this module */
 int test_munmap_sigsegv(void) {
```

Comment:
I think that with clever code shuffling you could get nice abstraction for signal handling and rewrite code to something like this:

```c
int test_munmap_sigsegv(void) {
  void *addr = mmap_anon_prw(NULL, 0x4000);
  munmap(addr, 0x4000);

  siginfo_t si;
  if (expect_signal(SIGSEGV, &si)) {
    /* Try to access freed memory. It should raise SIGSEGV */
    int data = *((volatile int *)(addr + 0x2000));
    (void)data;
  }

  /* Check if SIGSEGV was handled correctly */
  assert(got_signal(SIGSEGV));
  assert(si.si_addr == (addr + 0x2000));
  assert(si.si_code == SEGV_MAPERR);

  return 0;
}
```

Beware it can require some macrology around `expect_signal` since it should call sigsetjmp directly and not as subfunction. Please refer to https://gcc.gnu.org/onlinedocs/gcc-12.2.0/gcc/Statement-Exprs.html#Statement-Exprs

Reply:
Okay. I have prepared sth like that.